### PR TITLE
Propose YAML-LD document definition

### PR DESCRIPTION
**Pull request by anatoly-scherbakov**
In this PR, I'd like to propose a draft definition of a YAML-LD document, as well as a few minor changes.

## Questions

1. Where can one find a canonical description of YAML → JSON transformation? I cannot see that in YAML spec itself but it seems important for a kind of definition I'd like to propose.
2. What is the comparative role of **Core Requirements** section vs **Conformance** section? How does one decide where to put a particular normative statement? Perhaps there is a guidelines document?

Thank you!

***
<a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/pull/58.html" title="Last updated on Jul 25, 2022, 4:12 PM UTC (5510e23)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/58/7138fbf...5510e23.html" title="Last updated on Jul 25, 2022, 4:12 PM UTC (5510e23)">Diff</a>

### CODE REVIEW

File: spec/index.html [hunk 1]
```diff
@@ -235,7 +235,7 @@
   <body>
     <section id='abstract'>
       <p>
-        In recent years, YAML emerged as a more concise format
+        In recent years, [[YAML]] emerged as a more concise format
         to represent information serialized as JSON,
```
Comment (TallTed) [inline spec/index.html:239]:
```suggestion
        to represent information that had been serialized as JSON,
```

File: spec/index.html [hunk 2]
```diff
@@ -254,7 +254,7 @@
       <h2>Introduction</h2>
       <p>
         JSON-LD is a JSON-based format to serialize Linked Data.
-        In recent years, YAML emerged as a more concise format
+        In recent years, [[YAML]] emerged as a more concise format
         to represent information serialized as JSON,
```
Comment (TallTed) [inline spec/index.html]:
```suggestion
        to represent information that had been serialized as JSON,
```

File: spec/index.html [hunk 3]
```diff
@@ -253,25 +253,43 @@
...
+        to represent information serialized as [[JSON]],
         including API specification, data schemas, and Linked Data.
       </p>
       <p>
         Since YAML is more expressive than JSON,
         both in the available data types and in the document structure
         (see [[I-D.ietf-httpapi-yaml-mediatypes]]),
         this document identifies constraints on YAML
-        such that it can be used to represent JSON-LD documents.
+        such that any <a>YAML-LD document</a> can be represented in JSON-LD.
```
Comment (gkellogg) [inline spec/index.html:266]: This seems a non-sequitur: you discuss constraints on YAML that make YAML-LD can be represented as JSON-LD without first defining how YAML-LD relates to YAML. Maybe it needs to have a definition moved forward or referenced.

File: spec/index.html [hunk 4]
```diff
@@ -253,25 +253,43 @@
...
     </section>
 
   <section id="conformance">
-    <p>A <a>YAML-LD document</a> complies with this specification if ...</p>
-    <p class="ednote">Define <dfn>YAML-LD document</dfn> somewhere.</p>
+    <h2>YAML-LD stream</h2>
+
+    <p>
+      <dfn>YAML-LD stream</dfn> is a <a href="https://yaml.org/spec/1.2.2/#92-streams">YAML stream</a>
+      of <a>YAML-LD documents</a>.
```
Comment (gkellogg) [inline spec/index.html:275]: Definition is unreferenced. Maybe reference in an ednote, until text is added that uses the definition.

File: spec/index.html [hunk 5]
```diff
@@ -253,25 +253,43 @@
...
     <p>
-      The term media type is imported from [[RFC6838]].
+      Under a <dfn>YAML-LD document</dfn>, we understand any [[YAML]] document that, after a conversion to [[JSON]],
+      is a valid [[JSON-LD]] document and can be interpreted as [[LINKED-DATA]].
+    </p>
+
+    <h2>Terms</h2>
+
+    <p>
+      The term "media type" is imported from [[RFC6838]].
```
Comment (gkellogg) [inline spec/index.html:292]: These term references should probably be in the form `<a data-dfn="RFC6838#media-type">media type</a>`.

I've offered to redo this along the lines of the JSON-LD docs, so they can be `<dfn data-dfn="RFC6838#media-type">media type</dfn>`, which has the same effect but allows other term references to simply be <a>media type</a>, for example.

**Review by gkellogg — CHANGES_REQUESTED** [review]
This is great; a couple of suggested changes.

**Comment by gkellogg** [comment]
See PR #59 which updates this PR.

File: spec/index.html (hunk 2 above)
Comment (anatoly-scherbakov) [reply spec/index.html]: Applied, thank you!

File: spec/index.html (hunk 5 above)
Comment (anatoly-scherbakov) [reply spec/index.html]: I didn't know this form, thank you for adding these links.

File: spec/index.html (hunk 3 above)
Comment (anatoly-scherbakov) [reply spec/index.html:273]: Added an explanatory paragraph, please let me know what you think.

File: spec/index.html (hunk 4 above)
Comment (anatoly-scherbakov) [reply spec/index.html:282]: Added a reference.

File: spec/index.html [hunk 6]
```diff
@@ -253,25 +253,43 @@
     <section>
       <h2>Introduction</h2>
       <p>
-        JSON-LD is a JSON-based format to serialize Linked Data.
-        In recent years, YAML emerged as a more concise format
-        to represent information serialized as JSON,
+        [[JSON-LD]] is a JSON-based format to serialize Linked Data.
+        In recent years, [[YAML]] emerged as a more concise format
+        to represent information serialized as [[JSON]],
```
Comment (TallTed) [inline spec/index.html:258]:
```suggestion
        to represent information that had previously been serialized as [[JSON]],
```

File: spec/index.html [hunk 7]
```diff
@@ -253,44 +253,99 @@
...
         Since YAML is more expressive than JSON,
         both in the available data types and in the document structure
         (see [[I-D.ietf-httpapi-yaml-mediatypes]]),
         this document identifies constraints on YAML
-        such that it can be used to represent JSON-LD documents.
+        such that any <a>YAML-LD document</a> can be represented in JSON-LD.
       </p>
     </section>
 
   <section id="conformance">
```
Comment (TallTed) [inline spec/index.html:277]: It seems like there should be some additional section breaks, at least at each H2 here (which I think should actually be H3, which will be automagically handled by Respec if the `<section>` tags are inserted appropriately), which appear to be sub-sections of `<section id="conformance">` which is IDed but not then `<hN>` titled `Conformance`....

Comment (gkellogg) [reply spec/index.html:277]: Our standard is to use `<section>` around every block which starts with `<Hn>`.

**Comment by gkellogg** [comment]
This PR was discussed in [today's meeting](https://json-ld.org/minutes/2022-07-20/).

File: spec/index.html (hunk 7 above)
Comment (gkellogg) [reply spec/index.html]: I updated to use sections everywhere.

**Comment by gkellogg** [comment]
@anatoly-scherbakov I ended up moving the term definitions into the introduction, out of conformance. I also merged the sections you created for `YAML-LD document` and `YAML-LD stream` into the general term definitions, as there wasn't enough there to warrant their own sections. And, anything more probably should go out of the introduction and into something like basic concepts.

**Review by gkellogg — APPROVED** [review]

File: spec/index.html [hunk 8]
```diff
@@ -232,65 +232,122 @@
...
-
-    <section id='sotd'>
-      <p>
-        This is an unofficial proposal.
-      </p>
-    </section>
+<body>
+  <section id='abstract'>
+    <p>
+      In recent years, [[YAML]] emerged as a more concise format
```
Comment (TallTed) [inline spec/index.html:238]:
```suggestion
      In recent years, [[YAML]] has emerged as a more concise format
```

File: spec/index.html [hunk 9]
```diff
@@ -232,65 +232,122 @@
...
 
-  <section id="conformance">
-    <p>A <a>YAML-LD document</a> complies with this specification if ...</p>
-    <p class="ednote">Define <dfn>YAML-LD document</dfn> somewhere.</p>
+  <section>
+    <h2>Introduction</h2>
     <p>
-      The term media type is imported from [[RFC6838]].
+      [[JSON-LD]] is a JSON-based format to serialize Linked Data.
+      In recent years, [[YAML]] emerged as a more concise format
```
Comment (TallTed) [inline spec/index.html:256]:
```suggestion
      In recent years, [[YAML]] has emerged as a more concise format
```

File: spec/index.html [hunk 10]
```diff
@@ -232,65 +232,122 @@
...
+      to represent information that had previously been serialized as [[JSON]],
+      including API specification, data schemas, and Linked Data.
```
Comment (TallTed) [inline spec/index.html:258]:
```suggestion
      including API specifications, data schemas, and Linked Data.
```

File: spec/index.html [hunk 11]
```diff
@@ -232,65 +232,122 @@
...
+      of <a>YAML-LD documents</a>.
+    </p>
+
+    <p class="note" title="Documents in the stream are independent of each other">
+        Each document in a stream has its own context, [[YAML]] directives, named anchors, and so on.
     </p>
+
+    <p>A <dfn>YAML-LD document</dfn> is any <a>YAML document</a> that,
+      after a conversion to [[JSON]],
+      is a valid <a>JSON-LD document</a> and can be interpreted as [[LINKED-DATA]].
```
Comment (TallTed) [inline spec/index.html:291]:
```suggestion
    <p>A <dfn>YAML-LD document</dfn> is any <a>YAML document</a> from
      which a conversion to [[JSON]] produces
      a valid <a>JSON-LD document</a> which can be interpreted as [[LINKED-DATA]].
```

File: spec/index.html [hunk 12]
```diff
@@ -318,29 +375,38 @@ <h2>Introduction</h2>
...
+    statements for documents are often phrased as statements on the properties of the document.</p>
+</section>
 
   <section id="basic-concepts" class="informative">
     <h2>Basic Concepts</h2>
 
     <p>
-      To ease writing and collaborating on JSON-LD documents, it is a common practice
-      to serialize them as YAML.
+      To ease writing and collaborating on [[JSON-LD]] documents, it is a common practice
```
Comment (TallTed) [inline spec/index.html:392]:
```suggestion
      To ease writing and collaborating on [[JSON-LD]] documents, it is becoming common practice
```

File: spec/index.html [hunk 13]
```diff
@@ -318,29 +375,38 @@ <h2>Introduction</h2>
...
+      <li>YAML supports different encodings, including UTF-8, UTF-16, and UTF-32;</li>
+      <li>YAML supports more native data types than JSON;</li>
       <li>the structure of a YAML document &mdash; that is, a named YAML representation graph &mdash;
-      is a rooted, directed graph that can have cycles.</li>
-      <li>YAML has the concept of stream, which is a sequence of documents.
+      is a rooted, directed graph that can have cycles;</li>
+      <li>YAML has the concept of a <a>YAML stream</a>, which is a sequence of documents.
           While a stream usually contains one document,
           streams with multiple documents are used to aggregate multiple,
           related, documents into a single file or network stream.
```
Comment (TallTed) [inline spec/index.html:412]:
```suggestion
      <li>YAML supports different encodings, including UTF-8, UTF-16, and UTF-32.</li>
      <li>YAML supports more native data types than JSON.</li>
      <li>the structure of a YAML document &mdash; that is, a named YAML representation graph &mdash;
      is a rooted, directed graph that can have cycles.</li>
      <li>YAML has the concept of a <a>YAML stream</a>, which is a sequence of documents.
          While a stream often contains one document,
          streams with multiple documents can be used to aggregate multiple,
          related, documents into a single file or network stream.
```

File: spec/index.html [hunk 14]
```diff
@@ -393,132 +459,164 @@ <h2>Basic Concepts</h2>
...
+      A <a>YAML-LD document</a> MAY contain named anchors and alias nodes,
       but its representation graph MUST NOT contain cycles.
       When interpreting the document as JSON-LD,
       alias nodes MUST be resolved by value to their target nodes.
     </p>
 
-    <div data-format="markdown">
-      Example: The following YAML-LD document
+    <p>
+      The <a>YAML-LD document</a> at the following example
```
Comment (TallTed) [inline spec/index.html:519]:
```suggestion
      The <a>YAML-LD document</a> in the following example
```

File: spec/index.html (hunk 13 above)
Comment (gkellogg) [reply spec/index.html:412]: It's not allowing me to commit this suggestion; I'll do it via direct push.

File: spec/index.html (hunk 10 above)
Comment (TallTed) [reply spec/index.html]: "API specification" is singular; "data schemas" and "Linked Data" are plural.  I don't think it makes sense to change the plurals to singulars, hence my suggested change.

Comment (TallTed) [reply spec/index.html]: @gkellogg -- "API specification" is singular; "data schemas" and "Linked Data" are plural.  I don't think it makes sense to change the plurals to singulars, hence my suggested change.

**Pull request merged**